channel/channel.go

```diff
@@ -20,34 +20,33 @@ func NewChannel(bufferSize int) *Channel {
 		closed:     false,
 	}
 
-	ch.senderCond = sync.NewCond(&ch.senderMutex)
-	ch.receiverCond = sync.NewCond(&ch.receiverMutex)
+	ch.Cond = sync.NewCond(&ch.Mutex)
 	return ch
 }
 
 func (ch *Channel) Send(value interface{}) {
-	ch.senderMutex.Lock()
-	defer ch.senderMutex.Unlock()
+	ch.Mutex.Lock()
+	defer ch.Mutex.Unlock()
 
 	for len(ch.buffer) == ch.bufferSize && !ch.closed {
-		ch.senderCond.Wait()
+		ch.Cond.Wait()
 	}
 
 	if ch.closed {
 		panic("Sending on closed channel")
 	}
 
 	ch.buffer = append(ch.buffer, value)
-	ch.receiverCond.Signal()
+	ch.Cond.Signal()
```

Comment:
try update update your use of mutex and cond

```go
	c := sync.NewCond(&sync.Mutex{})
	c.L.Lock()
	for conditionTrue() == false {
		c.Wait()
	}
	c.L.Unlock()
```

Reply:
This will blocked the goroutine. I think, I already block the buffer